src/potentials.h

```diff
@@ -8,27 +8,138 @@
 #include "functionparser.h"
 #include "multipole.h"
 #include <array>
+#include <functional>
+#include "spdlog/spdlog.h"
+#include "spdlog/sinks/null_sink.h"
 
 namespace Faunus {
 namespace Potential {
 
 using namespace std::string_literals;
 
+//! type of a matrix containing pair potential coefficients
+typedef Eigen::MatrixXd TPairMatrix;
+typedef std::shared_ptr<TPairMatrix> TPairMatrixPtr;
+//! type of a function extracting a potential coefficient from the AtomData, e.g., sigma or eps
+typedef std::function<double(const AtomData&)> TExtractorFunc;
+//! type of a function defining a combination rule of a heterogeneous pair interaction
+typedef std::function<double(double, double)> TCombinatorFunc;
+//! type of a function modifying combinator's output
+typedef std::function<double(double)> TModifierFunc;
+
+/**
+ * @brief Data for a custom (heterogeneous) interaction between two given atom types.
+ *
+ * The very same format is used as for a homogeneous interaction specified directly on an atom type.
+ */
+struct InteractionData {
+    std::array<AtomData::Tid, 2> atom_id;
+    AtomData interaction;
+};
+
+void from_json(const json &j, std::vector<InteractionData> &interactions);
+void to_json(json &j, const std::vector<InteractionData> &interactions);
+
+/**
+ * @brief Known named combination rules for parameters of pair potential interaction.
+ *
+ * When adding a new one, add a json mapping. Also consider appending the PairMixer::getCombinator()
+ * method to recognize the new rule.
+ */
+enum ECombinationRule {COMB_UNDEFINED = 0, COMB_ARITHMETIC, COMB_GEOMETRIC, COMB_LORENTZ_BERTHELOT};
+NLOHMANN_JSON_SERIALIZE_ENUM( ECombinationRule, {
+    {COMB_UNDEFINED, "undefined"},
+    {COMB_ARITHMETIC, "arithmetic"},
+    {COMB_GEOMETRIC, "geometric"},
+    {COMB_LORENTZ_BERTHELOT, "lorentz_berthelot"},
+    {COMB_LORENTZ_BERTHELOT, "LB"}, // alternative non-canonical name
+})
+
+/**
+ * @brief Exception for handling pair potential initialization.
+ */
+struct PairPotentialException : public std::runtime_error {
+    PairPotentialException(const std::string msg) : std::runtime_error (msg) {};
```

Comment:
add inline or move to .cpp

Reply:
Constructors with an initializer list and empty body are used a lot in the header file. I can either move them all to cpp file or keep them all in the header file. The inline keyword is actually redundant as definition within the class makes the constructor inlined automatically. Suggestion?